### ISSUE TITLE
Apparently WASI start and WebAssembly start are two seperate functions

### ISSUE THREAD

**Issue by jtenner** [AssemblyScript/assemblyscript#1693]
I just want to open a discussion to get everything all squared away, and come to a consensus to aid in my development of the lunatic lib.  I have found myself confused and misunderstood, mostly because my understanding of WebAssembly is enough to simply get me in trouble 😂 

I have come to learn that each module when instantiated should definitely initialize memory properly, or there will be undefined behaviors. This means calling what I will refer to as the `WebAssembly.Start` method, which conflates with the apparently different `WASI._start()` function. Apparently these are different.

Of course, I'm going to advocate for lunatic here, and I can mostly describe how lunatic will work. When lunatic spawns a thread, it calls the `WebAssembly.Start`. If `ASC_WASI` is defined, it leaves memory uninitialzed, and plenty of undefined behavior starts to occur. So I've have to make some changes to make my tests pass as a result, and they feel wrong. Take the following asconfig as an example:

```ts
// asconfig.json
{
  "targets": {
    "debug": {
      "debug": true
    }
  },
  "options": {
     // I personally don't like to redefine these, because the "wasi" lib is really nice :)
    "use": ["abort=abort","trace=trace","seed=seed"],
    "lib": ["assembly/"],
    "exportTable": true
  }
}
```

I have to manually define an abort (which I have chosen to copy from std/assembly/wasi) and trace/seed etc. I also have to export a `_start()` function manually from the module entry point. This doesn't make sense from an end developer point of view, of course most of these details will be abstracted away by extending our default asconfig, so maybe this isn't so bad. 

If I might suggest a more elegant solution to this issue, I might suggest:

1. Memory and setup stuff should end up in WebAssembly.Start. We can tell it's memory and setup functions if they exist in `lib[]` entry points. EG, everything in `rt/*` and also everything provided as a `lib` entry point.
2. If `ASC_WASI` is defined, any top level statements in module entry points could end up in `WASI._start()` instead of `WebAssembly.Start`, and this will enable developers to ensure that `WASI` calls their start methods once, at least in spirit.

As far as I can tell, this might actually be expected behavior. Personally, I am mostly unsure how other CLI tools handle the different methods.

In the meantime, I would love to return back to doing this:

```ts
import "wasi";
```

Instead of defining abort, seed, and trace manually and doing this:

```ts
// Define all the things!
@global function abort(...): void {}
export function _start(): void {}
```

Given my current available workaround, I don't mind if nothing comes of this discussion. So there's nothing really game breaking for my development, and these issues can be documented on the `as-lunatic` package. I'm just curious if my suggested idea makes sense?

Thanks for everyone's help.

**Comment by dcodeIO** [comment]
I agree that we need a better story there, the current behavior doesn't seem right. Would suggest to remove the special `import "wasi";` behavior, so one can control generation of `_start` by either providing `--explicitStart` to achieve the same as the compiler currently does, or not use `--explicitStart` to have the option to emit both an implicit and a manual explicit start akin to a C `main` function defined in the entry file.

**Comment by jtenner** [comment]
Ideas I have.

```ts
asc entry.ts --explicitStart=symbol
asc entry.ts --use ASC_WASI_ENTRY=symbol

export function _start(): void {}

// potential decorator?
@wasiStart export function main(): void {}
```

Again, per my suggestion, any user entry files could be attached to the `_start()` method, and lib entry statements could be attached to the WebAssembly.Start method.

I *definitely* want this:

```ts
import "wasi";
```

**Comment by dcodeIO** [comment]
I am not suggesting to remove `import "wasi";`, just that it doesn't imply `--explicitStart` anymore.

**Comment by jtenner** [comment]
Ah. I misunderstood. This is just fine. It's the smallest change and produces the largest results.
